hub/core/storage/lru_cache.py

```diff
@@ -0,0 +1,163 @@
+from collections import OrderedDict
+from hub.core.storage.provider import StorageProvider
+from typing import Set
+
+# TODO use lock for multiprocessing
+class LRUCache(StorageProvider):
+    """LRU Cache that uses StorageProvider for caching"""
+
+    def __init__(
+        self,
+        cache_storage: StorageProvider,
+        next_storage: StorageProvider,
+        cache_size: int,
+    ):
+        """Initializes the LRUCache. This cache can be chained with
+
+        Args:
+            cache_storage (StorageProvider): The storage being used as the caching layer of the cache the.
+                This should be a base provider such as MemoryProvider, LocalProvider or S3Provider but not another LRUCache.
+            next_storage (StorageProvider): The next storage layer of the cache.
+                This can either be a base provider (i.e. it is the final storage) or another LRUCache (i.e. in case of chained cache).
+                While reading data, all misses from cache would be retrieved from here.
+                While writing data, the data will be written to the next_storage when cache_storage is full or flush is called.
```

Comment:
https://stackoverflow.com/questions/27087912/write-back-vs-write-through-caching

I don't think we've done our due diligence on considering write-through caching. We should at least make a point to do some benchmarking and compare the results. Write-through simplifies the API side considerably

Reply:
resolved for now after our discussion on slack. can be picked up again sometime later.